utility/transaction.go

```diff
@@ -42,112 +44,118 @@ func (u *UtilityContext) CheckTransaction(transactionProtoBytes []byte) error {
 		if totalSizeInBytes >= maxTransactionBytes {
 			// Add back popped transaction to be applied in a future block
 			err := u.Mempool.AddTransaction(txBytes)
 			if err != nil {
-				return nil, err
+				return nil, nil, err
 			}
 			break // we've reached our max
 		}
-		err = u.ApplyTransaction(transaction)
+		txResult, err := u.ApplyTransaction(index, transaction)
 		if err != nil {
 			// TODO: Properly implement 'unhappy path' for save points
 			if err := u.RevertLastSavePoint(); err != nil {
-				return nil, err
+				return nil, nil, err
 			}
 			totalSizeInBytes -= txSizeInBytes
+			continue
 		}
 		transactions = append(transactions, txBytes)
+		txResults = append(txResults, txResult)
+		index++
 	}
 	if err := u.EndBlock(proposer); err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return transactions, nil
+	return transactions, txResults, nil
 }
 
 // CLEANUP: Exposed for testing purposes only
-func (u *UtilityContext) AnteHandleMessage(tx *typesUtil.Transaction) (typesUtil.Message, typesUtil.Error) {
-	msg, err := tx.Message()
+func (u *UtilityContext) AnteHandleMessage(tx *typesUtil.Transaction) (msg typesUtil.Message, signer string, err typesUtil.Error) {
+	msg, err = tx.Message()
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	fee, err := u.GetFee(msg, msg.GetActorType())
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	pubKey, er := crypto.NewPublicKeyFromBytes(tx.Signature.PublicKey)
 	if er != nil {
-		return nil, typesUtil.ErrNewPublicKeyFromBytes(er)
+		return nil, "", typesUtil.ErrNewPublicKeyFromBytes(er)
 	}
 	address := pubKey.Address()
 	accountAmount, err := u.GetAccountAmount(address)
 	if err != nil {
-		return nil, typesUtil.ErrGetAccountAmount(err)
+		return nil, "", typesUtil.ErrGetAccountAmount(err)
 	}
 	accountAmount.Sub(accountAmount, fee)
 	if accountAmount.Sign() == -1 {
-		return nil, typesUtil.ErrInsufficientAmount(address.String())
+		return nil, "", typesUtil.ErrInsufficientAmount(address.String())
 	}
 	signerCandidates, err := u.GetSignerCandidates(msg)
 	if err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	var isValidSigner bool
 	for _, candidate := range signerCandidates {
 		if bytes.Equal(candidate, address) {
 			isValidSigner = true
+			signer = hex.EncodeToString(candidate)
 			break
 		}
 	}
 	if !isValidSigner {
-		return nil, typesUtil.ErrInvalidSigner()
+		return nil, signer, typesUtil.ErrInvalidSigner()
 	}
 	if err := u.SetAccountAmount(address, accountAmount); err != nil {
-		return nil, err
+		return nil, signer, err
 	}
 	if err := u.AddPoolAmount(typesUtil.PoolNames_FeeCollector.String(), fee); err != nil {
-		return nil, err
+		return nil, "", err
 	}
 	msg.SetSigner(address)
-	return msg, nil
+	return msg, signer, nil
 }
 
-func (u *UtilityContext) HandleMessage(msg typesUtil.Message) typesUtil.Error {
+func (u *UtilityContext) HandleMessage(msg typesUtil.Message) (err typesUtil.Error, messageType, recipient string) {
```

Comment:
Since we just copy-pasted `x.GetMessageName(), x.GetMessageRecipient()` on every line, we can simplify the method signature and just call `x.GetMessageName(), x.GetMessageRecipient()` where we invoke this method.

E.g.

Change

```go
	err, messageType, recipient := u.HandleMessage(msg)
	return tx.ToTxResult(u.LatestHeight, index, signer, recipient, messageType, err)
```

To something like:

```go
	err := u.HandleMessage(msg)
	return tx.ToTxResult(u.LatestHeight, index, signer, x.GetMessageRecipient(), x.GetMessageName(), err)
```

Reply:
Good point - will change